storage/cleanup.go

```diff
@@ -0,0 +1,159 @@
+		fm:                      fm,
+		filesetFilesBeforeFn:    fs.FilesetFilesBefore,
+		commitLogFilesBeforeFn:  fs.CommitLogFilesBefore,
+		commitLogFilesForTimeFn: fs.CommitLogFilesForTime,
+		deleteFilesFn:           fs.DeleteFiles,
+	}
+}
+
+func (mgr *cleanupManager) Cleanup(t time.Time) error {
+	multiErr := xerrors.NewMultiError()
+	filesetFilesStart := mgr.fm.FlushTimeStart(t)
+	if err := mgr.cleanupFilesetFiles(filesetFilesStart); err != nil {
+		detailedErr := fmt.Errorf("encountered errors when cleaning up fileset files for %v: %v", filesetFilesStart, err)
+		multiErr = multiErr.Add(detailedErr)
+	}
+	commitLogStart, commitLogTimes := mgr.commitLogTimes(t)
+	if err := mgr.cleanupCommitLogs(commitLogStart, commitLogTimes); err != nil {
+		detailedErr := fmt.Errorf("encountered errors when cleaning up commit logs for commitLogStart %v commitLogTimes %v: %v", commitLogStart, commitLogTimes, err)
+		multiErr = multiErr.Add(detailedErr)
+	}
+	return multiErr.FinalError()
+}
+
+func (mgr *cleanupManager) cleanupFilesetFiles(earliestToRetain time.Time) error {
+	multiErr := xerrors.NewMultiError()
+	shards := mgr.database.getOwnedShards()
+	for _, shard := range shards {
+		shardID := shard.ShardNum()
+		expired, err := mgr.filesetFilesBeforeFn(mgr.filePathPrefix, shardID, earliestToRetain)
+		if err != nil {
+			detailedErr := fmt.Errorf("encountered errors when getting fileset files for prefix %s shard %d: %v", mgr.filePathPrefix, shardID, err)
+			multiErr = multiErr.Add(detailedErr)
+		}
+		if err := mgr.deleteFilesFn(expired); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	}
+
+	return multiErr.FinalError()
+}
+
+// NB(xichen): since each commit log contains data needed for bootstrapping not only
+// its own block size period but also its left and right block neighbors due to past
+// writes and future writes, we need to shift flush time range by block size as the
+// time range for commit log files we need to check.
+func (mgr *cleanupManager) commitLogTimeRange(t time.Time) (time.Time, time.Time) {
+	flushStart, flushEnd := mgr.fm.FlushTimeStart(t), mgr.fm.FlushTimeEnd(t)
+	return flushStart.Add(-mgr.blockSize), flushEnd.Add(-mgr.blockSize)
+}
+
+// commitLogTimes returns the earliest time before which the commit logs are expired,
+// as well as a list of times we need to clean up commit log files for.
+func (mgr *cleanupManager) commitLogTimes(t time.Time) (time.Time, []time.Time) {
+	earliest, latest := mgr.commitLogTimeRange(t)
+
+	// TODO(xichen): preallocate the slice here
+	var commitLogTimes []time.Time
+	for commitLogTime := latest; !commitLogTime.Before(earliest); commitLogTime = commitLogTime.Add(-mgr.blockSize) {
+		hasFlushedAll := true
+		leftBlockStart := commitLogTime.Add(-mgr.blockSize)
+		rightBlockStart := commitLogTime.Add(mgr.blockSize)
+		for blockStart := leftBlockStart; !blockStart.After(rightBlockStart); blockStart = blockStart.Add(mgr.blockSize) {
+			if !mgr.fm.HasFlushed(blockStart) {
```

Comment:
It'll eventually get cleaned up once the data expires, i.e., when its timestamp is earlier than now - retentionPeriod.